fedimint-client/src/db.rs

```diff
@@ -557,3 +557,58 @@ pub async fn remove_old_and_persist_new_inactive_states(
         dbtx.insert_new_entry(&state, &inactive_state).await;
     }
 }
+
+/// Migrates a particular state by looping over all active and inactive states.
+/// If the `migrate` closure returns `None`, this state was not migrated and
+/// should be added to the new state machine vectors.
+pub async fn migrate_state<S: State + IntoDynInstance<DynType = DynState>>(
+    module_instance_id: ModuleInstanceId,
+    active_states: Vec<(Vec<u8>, OperationId)>,
+    inactive_states: Vec<(Vec<u8>, OperationId)>,
+    decoders: ModuleDecoderRegistry,
+    migrate: fn(
+        &[u8],
+        ModuleInstanceId,
+        &ModuleDecoderRegistry,
+    ) -> anyhow::Result<Option<DynState>>,
```

Comment:
* Why return a `Result`, what do we do if anything fails? Imo the developer should decide what to do in that case (delete state?)
* TODO(follow-up): not use `DynState` and just return `Vec<u8>`, then we can also get rid of `ModuleInstanceId`

Reply:
Returning an error might let us panic nicer (with more information about which module, etc.) in a higher level.

Reply:
Follow up: https://github.com/fedimint/fedimint/issues/4823